### PR TITLE
[GEOS-10696] Allow configuration of Output Format types allowed in GetFeature (2.22.x)

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-10696](https://osgeo-org.atlassian.net/browse/GEOS-10696)

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: doc/en/user/source/services/wfs/webadmin.rst [hunk 1]
```diff
@@ -165,4 +176,4 @@ Select default language for WFS Service.
    
    Default language
 
-See :ref:`internationalization` section for a how this setting is used.
\ No newline at end of file
+See :ref:`internationalization` section for a how this setting is used.
```
Comment (jodygarnett) [inline doc/en/user/source/services/wfs/webadmin.rst:179]: > See internationalization section for how this setting is used

File: doc/en/user/source/services/wfs/webadmin.rst [hunk 2]
```diff
@@ -66,6 +66,17 @@ A list of values to be declared in all feature types can be provided in the WFS
 
 The list will be used only for the capabilities document generation. It does not limit the actual target SRS usage in GetFeature requests.
 
+Configuration of Output Format types allowed in GetFeature
+----------------------------------------------------------
+
+Checking the ``Enable Output Format type checking`` checkbox will enable restrictions on the available output formats for GetFeature requests.  It will also limit what Output Format types will be displayed as available in GetCapabilities responses. GetFeature requests with Output types not included in the ``Allowed Output types`` panel will result in an ``Invalid Parameter ServiceException`` response.
```
Comment (jodygarnett) [inline doc/en/user/source/services/wfs/webadmin.rst:72]: Please use ``guidable`` for user interface elements:
```
:guilabel:`Enable output format type checking`
```

File: doc/en/user/source/services/wfs/webadmin.rst [hunk 3]
```diff
@@ -66,6 +66,17 @@ A list of values to be declared in all feature types can be provided in the WFS
 
 The list will be used only for the capabilities document generation. It does not limit the actual target SRS usage in GetFeature requests.
 
+Configuration of Output Format types allowed in GetFeature
```
Comment (jodygarnett) [inline doc/en/user/source/services/wfs/webadmin.rst:69]: Capitalization of section names:

> Configuration of output format types allowed in GetFeature

- https://docs.geoserver.org/latest/en/docguide/style.html#capitalization-of-page-names

File: src/web/wfs/src/main/java/org/geoserver/wfs/web/OutputTypesFormComponent.html [hunk 4]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
+  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml"
+	xmlns:wicket="http://wicket.apache.org/">
+<body>
+	<wicket:panel>
```
Comment (jodygarnett) [inline src/web/wfs/src/main/java/org/geoserver/wfs/web/OutputTypesFormComponent.html:6]: tabs vs spaces; do we have a convention for html pages?

File: src/web/wfs/src/main/java/org/geoserver/wfs/web/OutputTypesFormComponent.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+import org.apache.wicket.extensions.markup.html.form.palette.component.Recorder;
+import org.apache.wicket.extensions.markup.html.form.palette.theme.DefaultTheme;
+import org.apache.wicket.markup.html.basic.Label;
+import org.apache.wicket.markup.html.form.FormComponentPanel;
+import org.apache.wicket.model.IModel;
+import org.apache.wicket.model.Model;
+import org.apache.wicket.model.ResourceModel;
+import org.geoserver.web.wicket.SimpleChoiceRenderer;
+
+public class OutputTypesFormComponent extends FormComponentPanel<String> {
```
Comment (jodygarnett) [inline src/web/wfs/src/main/java/org/geoserver/wfs/web/OutputTypesFormComponent.java:25]: Class javadoc

File: src/web/wfs/src/main/resources/GeoServerApplication.properties [hunk 6]
```diff
@@ -26,6 +26,10 @@ WFSAdminPage.shapeZipPrjFormat=Use ESRI WKT format for SHAPE\u002DZIP generated
 WFSAdminPage.includeWFSRequestDumpFile=Include WFS request dump file
 WFSAdminPage.storedQueries = Stored Queries
 WFSAdminPage.complexToSimple = Activate complex to simple features conversion
+WFSAdminPage.getFeatureOutputTypes = Allowed Output Format types for a GetFeature request
+OutputTypesFormComponent.outputTypeCheckingEnabled = Enable Output Format type checking
+OutputTypesFormComponent.selectedHeader  = Allowed Output types
+OutputTypesFormComponent.availableHeader = Available Output types
```
Comment (jodygarnett) [inline src/web/wfs/src/main/resources/GeoServerApplication.properties:32]:
```suggestion
WFSAdminPage.getFeatureOutputTypes = Allowed output formats for a GetFeature request
OutputTypesFormComponent.outputTypeCheckingEnabled = Enable output format type checking
OutputTypesFormComponent.selectedHeader  = Allowed output formats
OutputTypesFormComponent.availableHeader = Available output formats
```

File: src/wfs/src/main/java/org/geoserver/wfs/outputformat/WFSOutputFormatCallback.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        }
+        return operation;
+    }
+    /**
+     * Builds a ServiceException for an unallowed output format
+     *
+     * @param outputFormat the output format requested
+     * @return the ServiceException
+     */
+    public ServiceException InvalidParameterException(String outputFormat) {
```
Comment (jodygarnett) [inline src/wfs/src/main/java/org/geoserver/wfs/outputformat/WFSOutputFormatCallback.java:76]: This method does not seem to have value:
- Not as a public method?
- By having this method the resulting service exception is raised against the wrong method.

If you really want to document this idea make a subclass of ServiceException and what you need to do in the constructor:
```java
public class OutputFormatServiceException extends ServiceException {
    public OutputFormatServiceException(String output format){
       super( "Invalid Output Format Parameter " + outputFormat, INVALID_PARAMETER_VALUE);
    }
}
```

**Review by jodygarnett — CHANGES_REQUESTED** [review]
Some small feedback on capitalization and exception use.

Q: Is there some interaction here between these allowed output formats and the WFS Settings to manage output formats (example "Override GML 3.2 MIME type")?

File: src/wfs/src/main/java/org/geoserver/wfs/outputformat/WFSOutputFormatCallback.java (hunk 7 above)
Comment (turingtestfail) [reply src/wfs/src/main/java/org/geoserver/wfs/outputformat/WFSOutputFormatCallback.java:76]: Jody, this is a backport of changes that were recently merged into main.  How would you like me to proceed?  Is it ok that by making the requested changes, the 2.22.x would be out of synch with main?  Should I open a new PR against main to make these changes there first?

**Comment by jodygarnett** [comment]
Oh, I see I missed my chance to review the original.

Please proceed with the backport; have a look at the proposed changes and if you agree we can make another PR.

File: src/wfs/src/main/java/org/geoserver/wfs/outputformat/WFSOutputFormatCallback.java (hunk 7 above)
Comment (jodygarnett) [reply src/wfs/src/main/java/org/geoserver/wfs/outputformat/WFSOutputFormatCallback.java:76]: Oh, I see I missed my chance to review the original.

Please proceed with the backport; have a look at the proposed changes and if you agree we can make another PR.

**Pull request merged**